components/org.wso2.carbon.identity.application.authenticator.hypr.common/src/main/java/org/wso2/carbon/identity/application/authenticator/hypr/common/web/HYPRAuthorizationAPIClient.java

```diff
@@ -0,0 +1,207 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.application.authenticator.hypr.common.web;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.reflect.TypeToken;
+import org.apache.http.HttpEntity;
+import org.apache.http.HttpResponse;
+import org.apache.http.util.EntityUtils;
+import org.wso2.carbon.identity.application.authenticator.hypr.common.constants.HyprAuthenticatorConstants;
+import org.wso2.carbon.identity.application.authenticator.hypr.common.exception.HYPRAuthnFailedException;
+import org.wso2.carbon.identity.application.authenticator.hypr.common.exception.HYPRClientException;
+import org.wso2.carbon.identity.application.authenticator.hypr.common.model.DeviceAuthenticationRequest;
+import org.wso2.carbon.identity.application.authenticator.hypr.common.model.DeviceAuthenticationResponse;
+import org.wso2.carbon.identity.application.authenticator.hypr.common.model.RegisteredDevice;
+import org.wso2.carbon.identity.application.authenticator.hypr.common.model.RegisteredDevicesResponse;
+import org.wso2.carbon.identity.application.authenticator.hypr.common.model.StateResponse;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.security.NoSuchAlgorithmException;
+import java.util.List;
+
+/**
+ * The HYPRAuthorizationAPIClient class contains all the functions related to handling the API calls to the HYPR server.
+ **/
+public class HYPRAuthorizationAPIClient {
+
+    /**
+     * Call the HYPR server API to retrieve the registered devices.
+     *
+     * @param username The username provided by the user.
+     * @return response         A HTTPResponse object.
+     * @throws HYPRAuthnFailedException Exception throws when there is an error occurred when retrieving the
+     *                                  registered devices via the api call
+     */
+    public static RegisteredDevicesResponse getRegisteredDevicesRequest(String baseUrl, String appId, String apiToken,
+                                                                        String username)
+            throws HYPRAuthnFailedException {
+
+        // Device info URL : {{baseUrl}}/rp/ api/oob/client/authentication/{{appId}}/{{username}}/devices
+        String deviceInfoURL = String.format("%s%s%s/%s/devices", baseUrl,
+                HyprAuthenticatorConstants.HYPR.HYPR_USER_DEVICE_INFO_PATH, appId, username);
+
+        try {
+            HttpResponse response = HYPRWebUtils.httpGet(apiToken, deviceInfoURL);
+
+            if (response.getStatusLine().getStatusCode() == 401) {
+                throw getHyprAuthnFailedException(
+                        HyprAuthenticatorConstants.ErrorMessages.HYPR_ENDPOINT_API_TOKEN_INVALID_FAILURE);
+            } else if (response.getStatusLine().getStatusCode() > 400) {
```

Comment:
Don't we need to handle 400 bad requests?

Reply:
Even for an invalid request the API call doesn't return Bad request instead it returns an empty array. That's why I didn't specifically handle it.